### ISSUE TITLE
Can I draw the line when I run the UE4

### ISSUE THREAD

**Issue by EricJin101** [Microsoft/AirSim#1393]
![screenshot from 2018-09-13 17-14-55](https://user-images.githubusercontent.com/26600982/45479377-ee7e1900-b778-11e8-804c-6f1d89dba390.png)
As the picture shows...

**Comment by EricJin101** [comment]
I need to display the global path in UE, how can I make it?

**Comment by DavidLSmyth** [comment]
If you edit BP_FlyingPawn to include a draw debug line node you should get the result shown in the picture. Here is what I used in my project: 
![image](https://user-images.githubusercontent.com/15518585/45484362-c3c99b80-b74b-11e8-9558-5078a37cba42.png)

**Comment by EricJin101** [comment]
> If you edit BP_FlyingPawn to include a draw debug line node you should get the result shown in the picture. Here is what I used in my project:
> ![image](https://user-images.githubusercontent.com/15518585/45484362-c3c99b80-b74b-11e8-9558-5078a37cba42.png)

What I really wanted to show was that once I had the planned trajectory I could visualize it on the AirSim directly, similar to the way one can in rviz.

**Comment by DavidLSmyth** [comment]
Your question seems pretty vague. If you want to view a planned trajectory you could just run the simulation with the clock significantly sped up?  As far as I know there is no way to visualise a planned trajectory without running the game as the simulator will start when the game runs, and the simulator is what estimates the state of the vehicles (simulator code [here](https://github.com/Microsoft/AirSim/blob/a4fcf568e75b6d675d5d127754757d673ce33ecb/AirLib/include/vehicles/multirotor/firmwares/simple_flight/AirSimSimpleFlightEstimator.hpp))

**Comment by DavidLSmyth** [comment]
I updated the api to visualise planned waypoints. I made the changes to worldSimApi, similar to how the printLogMessage functionality was added. I can submit a pull request if this would be useful to add to the api.
![image](https://user-images.githubusercontent.com/15518585/47017957-7d50cd80-d14b-11e8-85a6-da33afe2b789.png)

**Comment by EricJin101** [comment]
thank you! it will be extremely useful! I believe it is the exactly API I want.

**Comment by EricJin101** [comment]
> I updated the api to visualise planned waypoints. I made the changes to worldSimApi, similar to how the printLogMessage functionality was added. I can submit a pull request if this would be useful to add to the api.
> ![image](https://user-images.githubusercontent.com/15518585/47017957-7d50cd80-d14b-11e8-85a6-da33afe2b789.png)

thanks a lot! It's really useful

**Comment by DavidLSmyth** [comment]
I have it in my fork: https://github.com/DavidLSmyth/AirSim/
Changes are:

- You can now draw debug lines relative to a vehicle or to the 0,0,0 coordinate in the environment
- There is an option to turn a debug tracing line on for the RAV which tracks its path. 

Examples can be found in takeoff.py, new methods are showPlannedWaypoints and simShowDebugLines.

It's slightly messy at the moment as I had some trouble figuring out how to use Vector3r with RPC but I will fix that soon.

**Comment by EricJin101** [comment]
> I have it in my fork: https://github.com/DavidLSmyth/AirSim/
> Changes are:
> 
>     * You can now draw debug lines relative to a vehicle or to the 0,0,0 coordinate in the environment
> 
>     * There is an option to turn a debug tracing line on for the RAV which tracks its path.
> 
> 
> Examples can be found in takeoff.py, new methods are showPlannedWaypoints and simShowDebugLines.
> 
> It's slightly messy at the moment as I had some trouble figuring out how to use Vector3r with RPC but I will fix that soon.

I got an error...
_/vehicles/multirotor/api/MultirotorApiBase.cpp:7:10: fatal error: 
      'DrawDebugHelpers.h' file not found
#include "DrawDebugHelpers.h"_

**Comment by DavidLSmyth** [comment]
What unreal engine version and OS are you using?

**Comment by EricJin101** [comment]
> What unreal engine version and OS are you using? It's included in multiple files that existed before I made changes so strange that it's only showing that error now
> ![image](https://user-images.githubusercontent.com/15518585/47209405-2edc4280-d388-11e8-9ef9-e18b9d802bd3.png)
> 
> I'm developing on windows 10, visual studio 2017, unreal engine 4.18 so there may be compatibility issues.

sorry for the reply is not timely, I'm developing on Ubuntu 16.04, and UE is 4.20, my computer broke down last weekend... 
I will try it again this week

**Comment by ZWYDeveloper** [comment]
> > I have it in my fork: https://github.com/DavidLSmyth/AirSim/
> > Changes are:
> > ```
> > * You can now draw debug lines relative to a vehicle or to the 0,0,0 coordinate in the environment
> > 
> > * There is an option to turn a debug tracing line on for the RAV which tracks its path.
> > ```
> > Examples can be found in takeoff.py, new methods are showPlannedWaypoints and simShowDebugLines.
> > It's slightly messy at the moment as I had some trouble figuring out how to use Vector3r with RPC but I will fix that soon.
> 
> I got an error...
> _/vehicles/multirotor/api/MultirotorApiBase.cpp:7:10: fatal error: 'DrawDebugHelpers.h' file not found #include "DrawDebugHelpers.h"_
Hello, I have encountered this problem, have you solved it?

**Comment by ZWYDeveloper** [comment]
Where is this file "drawdebughelpers.h" from?UE4?So how do you compile airsim?

**Comment by EricJin101** [comment]
@ZWYDeveloper I am on a business trip, sadly, I didn't figure it out yet..

**Comment by DavidLSmyth** [comment]
What compiler are you using? I have ubuntu dual partitioned so I'll try it and post a fix. I imagine it's an issue with the compiler. DrawDebugHelpers is from UE, provides debugging visualisation functionality

**Comment by ZWYDeveloper** [comment]
@DavidLSmyth Yes, I succeeded, I send this file before compilation airsim commented out, put the compiled plug-in ue4 compilation in the scene and then opens, the effect of this is I compiled, is nice, but I want to optimize the trajectory of the lines, I need to take a look at you to modify the API, thank you for your open source share, thank you!

**Comment by ZWYDeveloper** [comment]
![image](https://user-images.githubusercontent.com/44432279/47481175-5b4efd80-d865-11e8-9d10-dafdd9dbc984.png)
![tim 20181025142432](https://user-images.githubusercontent.com/44432279/47481202-6efa6400-d865-11e8-9764-a50a50f50b4d.png)

**Comment by DavidLSmyth** [comment]
> @DavidLSmyth Yes, I succeeded, I send this file before compilation airsim commented out, put the compiled plug-in ue4 compilation in the scene and then opens, the effect of this is I compiled, is nice, but I want to optimize the trajectory of the lines, I need to take a look at you to modify the API, thank you for your open source share, thank you!

The changes were pretty straightforward, they are all contained in one commit. Happy to help!

**Comment by RVBldr** [comment]
Internal VSTF 7413.

**Comment by AloshkaD** [comment]
@DavidLSmyth  Thank you for sharing your addition to path plotting. I had an error with`self.simShowPawnPath(True, 3, 8,vehicle_name)`
as follows 
`msgpackrpc.error.RPCError: rpclib: server could not find function 'simShowPawnPath' with argument count 4.`
I have installed your airsim fork to make sure I don't miss anything. 
Also do I still need to edit BP_FlyingPawn ?

**Comment by DavidLSmyth** [comment]
> @DavidLSmyth Thank you for sharing your addition to path plotting. I had an error with`self.simShowPawnPath(True, 3, 8,vehicle_name)`
> as follows
> `msgpackrpc.error.RPCError: rpclib: server could not find function 'simShowPawnPath' with argument count 4.`
> I have installed your airsim fork to make sure I don't miss anything.
> Also do I still need to edit BP_FlyingPawn ?

I made some changes recently that may have broken some things, I'll have a look and get back to you when I've figured out what's gone wrong. If you don't need some of the latest functionality you could use one of the older commits (try 129596f7e964d96936b46a54c89b44f6a36bf57c)

**Comment by AloshkaD** [comment]
Thanks @DavidLSmyth  I look forward to that. Regarding the commit that you referenced, it does not seem related. It's an update to the  simGetPositionWRTOrigin method.

**Comment by Hoshea7** [comment]
Hi @AloshkaD ,

I still cannot figure out how to draw the line of the trajectory of multirotor, could you kindly explain more how you complete this ?

**Comment by tteague19** [comment]
@DavidLSmyth I am encountering an error that "rpclib: server could not find function 'simShowDebugLines' with argument count 0" when I try to use your fork. Strangely, the function is defined in your .hpp and .cpp files with the appropriate number of arguments. Would you be able to help with this issue? Thanks!

**Comment by DavidLSmyth** [comment]
I'll see if I can find the time to submit this as a clean pull request. Haven't done any maintenance on this fork for years...

**Comment by tteague19** [comment]
Thanks, @DavidLSmyth! In case you are unable to make the fix, do you know if there has been a similar capability developed within AirSim? Essentially, I want to indicate a connection between two or more multirotors for the purposes of visualization.

**Comment by DavidLSmyth** [comment]
> Thanks, @DavidLSmyth! In case you are unable to make the fix, do you know if there has been a similar capability developed within AirSim? Essentially, I want to indicate a connection between two or more multirotors for the purposes of visualization.

I don't think so (you can turn on debug lines showing where the drone has flown, that's the closest thing I'm aware of) but you could put this together using a blueprint rather than making a change to the API (if you've seen images of other people doing this I'd guess they've just used a blueprint). I got pretty frustrated with blueprints in the end so I'd recommend to do this from the API if possible!

**Comment by DavidLSmyth** [comment]
> Thanks, @DavidLSmyth! In case you are unable to make the fix, do you know if there has been a similar capability developed within AirSim? Essentially, I want to indicate a connection between two or more multirotors for the purposes of visualization.

Was just going through it there, maybe [simPlotLineList](https://microsoft.github.io/AirSim/api_docs/html/#airsim.client.VehicleClient.simPlotLineList) is what you want? I wasn't aware this had been added!

**Comment by tteague19** [comment]
It is! By coincidence, I found that utility earlier this week, as well. Ideally, I would plot the lines throughout the simulation, but it suffices to use it between movements, which seems to be the only way to do so. If I become more proficient with AirSim, I will consider implementing something closer to what I have in mind. Thank you again for your help, @DavidLSmyth!
